### PR TITLE
Implement `load_grid`

### LINKED ISSUE #125: load_grid method/function
I implemented a `load_grid` static method on `pysgrid` and, as we discussed in APIRUS, I am working on implementing the same here.

Right now we have 4 ways of loading the grid.  The UGrid class methods `UGrid.from_ncfile` and `UGrid.from_nc_dataset` and the functions `load_grid_from_ncfilename` and `load_grid_from_ncfilename` and `load_grid_from_nc_dataset`.

My plan is to implement the function `load_grid_from_nc_dataset` as a class method named `load_grid` and make it take either a string or a `netCDF4.Dataset`.  This function already takes the `UGrid` class as its input, so it is natural to become a method of that class.  Then I will write a thin wrapper for that method named `load_grid` so people from do:

```python
from pyugrid import load_grid

ug = load_grid(nc_url)
```

Akin to `pysgrid`'s API.

Any thoughts?

### PR DESCRIPTION
**Pull request by ocefpaf**
Closes #125 

**THIS PR IS NOT READY YET.  DO NOT MERGE!**

This PR brings `pysgrid` and `pyugrid` APIs closer.

Ping: @rsignell-usgs, @kwilcox and @ChrisBarker-NOAA please comment!

### CODE REVIEW

File: test/test_write_read.py [hunk 1]
```diff
@@ -13,28 +13,29 @@
...
+    Test with faces, edges, but no `face_coordinates` or `edge_coordinates`.
 
     """
-
     with chdir(test_files):
         grid = two_triangles()
         grid.save_as_netcdf('2_triangles.nc')
         # Read it back in and check it out.
-        grid2 = UGrid.from_ncfile('2_triangles.nc')
+        with Dataset('2_triangles.nc') as nc:
```
Comment (ocefpaf) [inline test/test_write_read.py]: Note to self: with statement is safer when reading/saving netCDF files during the tests.

**Comment by ocefpaf** [comment]
@rhattersley I plan to move things around and re-factor `ugrid.py` (mostly `load_grid`), but if you find some time I would love your input here. In particular with respect to the API and how we can make it easy to integrate `pyugrid` and `iris`.

**Comment by ocefpaf** [comment]
Waiting for #120, #130, #131, and  #132  to be sorted out before continue this PR to avoid conflict headaches. (Ping @ChrisBarker-NOAA.)

File: pyugrid/ugrid.py [hunk 2]
```diff
@@ -116,68 +109,175 @@ def __init__(self,
...
-        create a UGrid object from a netcdf file name (or opendap url)
+        Loads UGrid object from a `netCDF4.Dataset` object or file name/URL,
+        adding the data to the passed-in grid object.
 
-        :param nc_url: the filename or OpenDap url you want to load
+        It will load the mesh specified, or look for the first one it finds if
+        none is specified.
+
+        :param nc: The grid data to be loaded up.
+        :type nc: `netCDF4.Dataset` object or netCDF/URL string
```
Comment (ChrisBarker-NOAA) [inline pyugrid/ugrid.py]: maybe:
 
`netCDF4.Dataset` object or netCDF filename or URL string.

File: pyugrid/ugrid.py [hunk 3]
```diff
@@ -116,68 +109,175 @@ def __init__(self,
...
+
+        :param mesh_name=None: name of the mesh to load.
+        :type mesh_name: string
 
-        :param mesh_name=None: the name of the mesh you want. If None, then
-                               you'll get the only mesh in the file. If there
-                               is more than one mesh in the file, a ValueError
-                               Will be raised
         :param load_data=False: flag to indicate whether you want to load the
```
Comment (ChrisBarker-NOAA) [inline pyugrid/ugrid.py]: Hmm -- if the netcdf Variables are lazy loaded, maybe we don't need this at all. When I first wrote this, UVars turned the netcdf variable objects into numpy arrays right off -- you don't want that, but if you just keep a reference to the variables in a UVAR, then why not have them all loaded at once to be queried?

File: pyugrid/ugrid.py [hunk 4]
```diff
@@ -599,11 +688,8 @@ def save_as_netcdf(self, filepath):
         """
         mesh_name = self.mesh_name
 
-        # FIXME: Why not use netCDF4.Dataset instead of renaming?
-        from netCDF4 import Dataset as ncDataset
         # Create a new netcdf file.
-        with ncDataset(filepath, mode="w", clobber=True) as nclocal:
-
+        with Dataset(filepath, mode='w', clobber=True) as nclocal:
```
Comment (ChrisBarker-NOAA) [inline pyugrid/ugrid.py]: we should probably have a way for the user to pass in options to netCDF4 - netcdf version, etc....

File: pyugrid/ugrid.py [hunk 5]
```diff
@@ -770,3 +856,34 @@ def save_as_netcdf(self, filepath):
...
+    :param load_data=False: flag to indicate whether you want to load the
+                            associated data or not. The mesh will be loaded
+                            in any case. If False, only the mesh will be
+                            loaded. If True, then all the data associated
+                            with the mesh will be loaded.
+                            This could be huge!
+    :type load_data: boolean
+
+    """
+    if isinstance(nc, Dataset):
```
Comment (ChrisBarker-NOAA) [inline pyugrid/ugrid.py]: why not:

```
if not isinstance(nc, Dataset):
    nc = Dataset(nc, 'r')
```

or even:

nc = Dataset(nc, 'r') if not isinstance(nc, Dataset)

**Comment by ChrisBarker-NOAA** [comment]
A few comments -- but this looks great to me.

**Comment by ChrisBarker-NOAA** [comment]
@ocefpaf wrote:
"""Waiting for #120, #130, #131, and #132 to be sorted out before continue this PR """

as it happens, it looks like this one has no conflicts, but the others all do. 

But I'm going to let you clean those up if you don't mind :-)

**Comment by ocefpaf** [comment]
Chris, thanks for the review! I'll make the changes and ping you back.

I am flying to Ocean Sciences now. I so won't work on this for a while.
There are still a lot to do here. Add more tests, ensure it works with
remote datasets, docs, etc...

On Feb 17, 2016 8:02 PM, "Chris Barker" <notifications@github.com> wrote:
>
> @ocefpaf wrote:
> """Waiting for #120, #130, #131, and #132 to be sorted out before
continue this PR """
>
> as it happens, it looks like this one has no conflicts, but the others
all do.

**Comment by ChrisBarker-NOAA** [comment]
Looks like there are tests that need updating, and conflicts to resolve. But once that's done, it looks good to me.

**Comment by ocefpaf** [comment]
@ChrisBarker-NOAA this PR is not ready.  I will ping you once I am done, but there are many problems with it yet.

**Comment by ChrisBarker-NOAA** [comment]
ping! -- it would be good to get this one cleaned up and in. I've lost track of the state.

**Comment by ocefpaf** [comment]
I'll re-start this from scratch at a later time. Too busy with conda-forge at the moment.

**Comment by ChrisBarker-NOAA** [comment]
NP -- thanks! great work on conda-forge, BTW!

**Pull request closed without merging**